plugins/MidiExport/MidiExport.cpp

```diff
@@ -216,27 +216,27 @@ bool MidiExport::tryExport(const TrackContainer::TrackList &tracks,
 				int len = n.toElement().attribute("steps", "1").toInt() * 12;
 
 				// for each pattern clip of the current pattern track (in song editor)
-				for (auto it = plist.begin(); it != plist.end(); ++it)
+				for (auto& pair : plist)
```

Comment:
Or maybe it's time for another

    for (auto&& [start, end] : plist)

Which I don't understand half of even after reading [Dom's explanation](https://discord.com/channels/203559236729438208/784594576223633478/1012474458956640339) but it looks cool and dangerous

Reply:
> Or maybe it's time for another
> 
> ```
> for (auto&& [start, end] : plist)
> ```
> 
> Which I don't understand half of even after reading [Dom's explanation](https://discord.com/channels/203559236729438208/784594576223633478/1012474458956640339) but it looks cool and dangerous

Yeah, I'm somewhat unsure as well, but I really think it depends on context and what the ``range-expression`` is (in this case ``plist``). I've been doing some reading, and from what I understand, ``auto&&`` (r-value references) only bind to r-values, ``auto&`` (l-value reference) only bind to l-values, and ``const auto&`` (const l-value reference) can bind to both l-values and r-values.

So, I'm guessing that the most correct choice of using ``auto&``, ``auto&&``, or ``const auto&`` depends on what kind of value ``range-expression`` returns. It's important to note that other value categories, such as glvalues and xvalues, fall under either an l-value or an r-value, so they apply as well.

In this case, if ``range-expression`` (``plist``) gives back an l-value, ``auto&`` or ``const auto&`` can be used. If ``range-expression`` gives back an r-value, ``auto&&`` or ``const auto&`` can be used. So if the expression is not changed given the range-based for loop (i.e it can be const), ``const auto &`` may be the most optimal choice given a wide variety of circumstances. 

However, it is possible to convert an l-value to an r-value given an implicit l-value to r-value conversion. Therefore, this would allow ``auto&&`` to bind to both l-values and r-lvaues.

So ultimately, I think the main benefit of using ``auto&&`` in range-based for loops is if you need to modify the value in question, but also need it to bind to both l-values and r-values.

Structured bindings don't really change this idea at all. The type of a structured binding is compiler generated (hence why we use ``auto`` for them), but at the end of the day that just means it's a ``range-declaration`` that has one is like any other declaration.

Given my current understanding of this, I think I will update my changes to use ``const auto&`` where applicable, and ``auto&&`` otherwise.

**TLDR: If you don't need to modify the range declaration, ``const auto&`` should be used. If you do need to modify the range declaration, ``auto&&`` should be used.**

Reply:
> from what I understand, `auto&&` (r-value references) only bind to r-values

It can bind to both `T&`(`auto` becomes `T&`) and `T&&`(`auto` becomes `T`).

> If you don't need to modify the range declaration, `const auto&` should be used. If you do need to modify the range declaration, `auto&&` should be used.

That sounds reasonable, indeed. I would use `T&` in case I need to both modify the declaration and make sure it's bound to an lvalue, BTW.

Reply:
> That sounds reasonable, indeed. I would use `T&` in case I need to both modify the declaration and make sure it's bound to an lvalue, BTW.

Yeah, I'll add that if the range-based for loop can go without needing to bind to r-values, a ``auto&`` is a nice choice mainly due to simplicity and readability reasons. 

...However, when seeing a range-based for loop, the compiler would generate something like this (from https://en.cppreference.com/w/cpp/language/range-for):

```cpp
    auto && __range = range-expression ;
    auto __begin = begin-expr ;
    auto __end = end-expr ;
    for ( ; __begin != __end; ++__begin)
    {

        range-declaration = *__begin;
        loop-statement

    }
```
And given ``range-declaration = *__begin;``, I read that the indirection operator ``*`` gives back an l-value, or more specifically a ``T&``. Then again, it can be overloaded, and in cases like ``std::vector<bool>``, dereferencing the result of ``std::vector<bool>::begin`` doesn't return an l-value, but a temporary r-value with type ``std::vector<bool>::reference``, as seen here (in particular with the GCC compiler).

```cpp
_GLIBCXX_NODISCARD _GLIBCXX20_CONSTEXPR
reference
operator*() const
{ return reference(_M_p, 1UL << _M_offset); }
 ```

Which cannot bind to an l-value reference, such as ``auto&``.

Okay, I think I got it now. As you stated, ``auto&`` should be used if it is sure to be bound to an l-value, which the indirection operator **usually** returns, but not always. This would then pave way for the use of ``auto&&``, but if you don't need to modify anything, ``const auto&`` works way better in terms of intent, simplicity, and readability.

Reply:
Good point for rare edge cases such as `std::vector<bool>`, which I never used.